File: modules/opd/src/opd/http_handlers/health_documents.py
"""Patient health document upload, list, and download (Azure Blob)."""

from __future__ import annotations

from math import ceil
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, Request, UploadFile
from fastapi.responses import RedirectResponse, Response

from opd.core.deps import DbSession, TenantId
from opd.core.principal import resolve_doctor_id
from opd.core.schemas_api import (
    HealthDocumentDownloadResponse,
    HealthDocumentListResponse,
    HealthDocumentSummary,
    HealthDocumentUploadResponse,
)
from opd.data_access.health_document_repo import HealthDocumentRepository
from opd.lib import azure_blob_storage
from opd.lib.file_upload_validation import (
    generate_health_document_path,
    sanitize_filename,
    validate_health_document_upload,
)

router = APIRouter(tags=["OpdHealthDocuments"])

UploaderId = Annotated[UUID, Depends(resolve_doctor_id)]


def _download_path(document_id: UUID) -> str:
    return f"/api/v1/opd/health-documents/{document_id}/download"


def _to_summary(row) -> HealthDocumentSummary:
    return HealthDocumentSummary(
        id=row.id,
        patient_id=row.patient_id,
        visit_id=row.visit_id,
        file_name=row.original_file_name,
        file_type=row.mime_type,
        document_title=row.document_title,
        hi_type=row.hi_type,
        uploaded_at=row.uploaded_at,
        download_url=_download_path(row.id),
    )


def _azure_unavailable(exc: Exception) -> HTTPException:
    return HTTPException(status_code=503, detail=str(exc))


def _accepts_json_sas_url(accept_header: str) -> bool:
    """JSON SAS URL only when the client did not ask for a file MIME type."""
    lowered = accept_header.lower()
    if "application/json" not in lowered:
        return False
    for binary in (
        "application/pdf",
        "application/octet-stream",
        "image/jpeg",
        "image/jpg",
        "image/png",
    ):
        if binary in lowered:
            return False
    return True


@router.post(
    "/patients/{patient_id}/health-documents",
    response_model=HealthDocumentUploadResponse,
    status_code=201,
)
async def upload_patient_health_document(
    patient_id: UUID,
    db: DbSession,
    tenant_id: TenantId,
    uploaded_by: UploaderId,
    file: UploadFile = File(...),
    hi_type: str = Form(...),
    document_title: str = Form(...),
    visit_id: UUID | None = Form(default=None),
) -> HealthDocumentUploadResponse:
    content = await file.read()
    mime_type = file.content_type or "application/octet-stream"
    try:
        validate_health_document_upload(mime_type, len(content))
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    folder_path = generate_health_document_path(patient_id, visit_id, hi_type)
    try:
        blob = azure_blob_storage.upload_health_document_blob(
            content,
            file.filename or "document",
            mime_type,
            folder_path,
        )
    except RuntimeError as exc:
        raise _azure_unavailable(exc) from exc

    repo = HealthDocumentRepository(db, tenant_id)
    row = repo.create(
        patient_id=patient_id,
        visit_id=visit_id,
        hi_type=hi_type.strip(),
        document_title=document_title.strip(),
        original_file_name=file.filename or "document",
        storage_key=blob.storage_key,
        blob_url=blob.blob_url,
        mime_type=mime_type,
        file_size_bytes=len(content),
        uploaded_by=uploaded_by,
    )
    db.commit()
    return HealthDocumentUploadResponse(
        id=row.id,
        patient_id=row.patient_id,
        visit_id=row.visit_id,
        file_name=row.original_file_name,
        file_type=row.mime_type,
        document_title=row.document_title,
        hi_type=row.hi_type,
        uploaded_at=row.uploaded_at,
    )


@router.get(
    "/patients/{patient_id}/health-documents",
    response_model=HealthDocumentListResponse,
)
def list_patient_health_documents(
    patient_id: UUID,
    db: DbSession,
    tenant_id: TenantId,
    visit_id: UUID | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=10, ge=1, le=100),
) -> HealthDocumentListResponse:
    repo = HealthDocumentRepository(db, tenant_id)
    rows, total = repo.list_for_patient(
        patient_id=patient_id,
        visit_id=visit_id,
        page=page,
        limit=limit,
    )
    db.commit()
    total_pages = max(1, ceil(total / limit)) if total else 0
    return HealthDocumentListResponse(
        success=True,
        count=len(rows),
        total=total,
        total_pages=total_pages,
        current_page=page,
        limit=limit,
        data=[_to_summary(row) for row in rows],
    )


@router.get(
    "/health-documents/{document_id}/download",
    response_model=HealthDocumentDownloadResponse,
    responses={
        200: {
            "description": "File bytes (when Accept is a file MIME type, not JSON-only)",
            "content": {"application/octet-stream": {}},
        },
        302: {"description": "Redirect to Azure Blob SAS URL (legacy clients)"},
    },
)
def download_health_document(
    document_id: UUID,
    request: Request,
    db: DbSession,
    tenant_id: TenantId,
) -> HealthDocumentDownloadResponse | RedirectResponse | Response:
    repo = HealthDocumentRepository(db, tenant_id)
    row = repo.get(document_id)
    if row is None or row.status != "active":
        raise HTTPException(status_code=404, detail="Document not found")

    accept = request.headers.get("accept", "")
    if _accepts_json_sas_url(accept):
        try:
            sas_url = azure_blob_storage.generate_blob_sas_url(
                row.blob_url,
                storage_key=row.storage_key,
                download_file_name=row.original_file_name,
            )
        except (RuntimeError, ValueError) as exc:
            raise _azure_unavailable(exc) from exc
        db.commit()
        return HealthDocumentDownloadResponse(url=sas_url)

    try:
        content, content_type = azure_blob_storage.download_health_document_bytes(row.storage_key)
    except RuntimeError as exc:
        raise _azure_unavailable(exc) from exc

    db.commit()
    if len(content) < 100:
        raise HTTPException(status_code=404, detail="Document file not found in storage")

    safe_name = sanitize_filename(row.original_file_name)
    media_type = row.mime_type or content_type
    return Response(
        content=content,
        media_type=media_type,
        headers={
            "Content-Disposition": f'attachment; filename="{safe_name}"',
            "Content-Length": str(len(content)),
        },
    )

File: services/web/src/features/create-rx/api/health-documents.ts
import { apiClient, apiClientBlob, apiClientFormData } from '@/lib/api-client';

const OPD_PREFIX = '/api/v1/opd';

export const HEALTH_DOCUMENT_HI_TYPES = [
  'Diagnostic Report Record',
  'Discharge Summary Record',
  'OP Consultation Record',
  'Prescription Record',
  'Immunization Record',
  'Invoice Record',
  'Wellness Record',
] as const;

export interface HealthDocumentSummary {
  id: string;
  patient_id: string;
  visit_id?: string | null;
  file_name: string;
  file_type: string;
  document_title: string;
  hi_type: string;
  uploaded_at: string;
  download_url: string;
}

export interface HealthDocumentListResponse {
  success: boolean;
  count: number;
  total: number;
  total_pages: number;
  current_page: number;
  limit: number;
  data: HealthDocumentSummary[];
}

export async function uploadHealthDocument(
  patientId: string,
  file: File,
  hiType: string,
  documentTitle: string,
  visitId?: string,
): Promise<HealthDocumentSummary> {
  const formData = new FormData();
  formData.append('file', file);
  formData.append('hi_type', hiType);
  formData.append('document_title', documentTitle);
  if (visitId?.trim()) {
    formData.append('visit_id', visitId.trim());
  }

  const uploaded = await apiClientFormData<HealthDocumentUploadResponse>(
    `${OPD_PREFIX}/patients/${patientId}/health-documents`,
    formData,
  );

  return {
    id: uploaded.id,
    patient_id: uploaded.patient_id,
    visit_id: uploaded.visit_id,
    file_name: uploaded.file_name,
    file_type: uploaded.file_type,
    document_title: uploaded.document_title,
    hi_type: uploaded.hi_type,
    uploaded_at: uploaded.uploaded_at,
    download_url: `${OPD_PREFIX}/health-documents/${uploaded.id}/download`,
  };
}

interface HealthDocumentUploadResponse {
  id: string;
  patient_id: string;
  visit_id?: string | null;
  file_name: string;
  file_type: string;
  document_title: string;
  hi_type: string;
  uploaded_at: string;
}

export async function fetchPatientHealthDocuments(
  patientId: string,
  visitId?: string,
  page = 1,
  limit = 50,
): Promise<HealthDocumentListResponse> {
  const params = new URLSearchParams({
    page: String(page),
    limit: String(limit),
  });
  if (visitId?.trim()) {
    params.set('visit_id', visitId.trim());
  }

  return apiClient<HealthDocumentListResponse>(
    `${OPD_PREFIX}/patients/${patientId}/health-documents?${params.toString()}`,
  );
}

function resolveDownloadApiPath(downloadPath: string): string {
  if (downloadPath.startsWith('http')) {
    try {
      return new URL(downloadPath).pathname;
    } catch {
      return downloadPath;
    }
  }
  if (downloadPath.startsWith('/')) {
    return downloadPath;
  }
  return `${OPD_PREFIX}/health-documents/${downloadPath}/download`;
}

function downloadAcceptHeader(fileName: string, mimeType?: string): string {
  const fromMime = mimeType?.trim();
  if (fromMime) return fromMime;
  const lower = fileName.toLowerCase();
  if (lower.endsWith('.pdf')) return 'application/pdf';
  if (lower.endsWith('.png')) return 'image/png';
  if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return 'image/jpeg';
  return 'application/octet-stream';
}

/** Stream file via OPD (same-origin; Azure blobs are not CORS-enabled for localhost). */
export async function downloadHealthDocument(
  downloadPath: string,
  fileName: string,
  mimeType?: string,
): Promise<void> {
  const apiPath = resolveDownloadApiPath(downloadPath);
  const accept = downloadAcceptHeader(fileName, mimeType);

  const blob = await apiClientBlob(apiPath, {
    headers: { Accept: accept },
  });

  if (blob.size < 100) {
    throw new Error('Downloaded file is empty or invalid. Try uploading again.');
  }

  const objectUrl = URL.createObjectURL(blob);
  const safeName = fileName.trim() || 'document';
  const link = document.createElement('a');
  link.href = objectUrl;
  link.download = safeName;
  link.rel = 'noopener';
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);

  if (accept === 'application/pdf' || safeName.toLowerCase().endsWith('.pdf')) {
    window.open(objectUrl, '_blank', 'noopener,noreferrer');
  }

  window.setTimeout(() => URL.revokeObjectURL(objectUrl), 60_000);
}
